#!/usr/bin/perl -w
use strict;

my $recruited = shift;		# the list of recruited reads, expected to be generated by HMM-GRASPx
my $out_dir = shift;		# the output directory
my $pep_seq = shift;		# the peptide sequence in FASTA format
my $nuc_seq = shift;		# the nucleotide sequence in FASTA format

my $list = "/usr/local/projdata/0599/projects/SPA/czhong/Works/OralMeta/Results/long_HMM.list";

if(!-e $out_dir)  {
  mkdir "$out_dir" or die "Cannot create directory";
}

open my $LIN, "<$list" or die "Cannot open file: $!\n";
while(<$LIN>)  {
  chomp;
  my @decom = split /\s+/, $_;
  if(!-e "$out_dir\/$decom[0]")  {
    mkdir "$out_dir/$decom[0]" or die "Cannot create directory";
  }
  # generating a list of IDs
  print "Generating the list of recruited reads for family $decom[0]...\n";
  system "cat $recruited \| grep \'$decom[0]\' \| awk \'\{print \$2\}\' >$out_dir\/$decom[0]\/id_list";
  if(!-s "$out_dir\/$decom[0]\/id_list")  {
    print "Empty file... Skip...\n";
    system "touch $out_dir\/$decom[0]\/pep.fasta";
    system "touch $out_dir\/$decom[0]\/nuc.fasta";
    next;
  }
  # extract the sequences
  print "Compiling the list of recruited peptide sequences...\n";
  system "perl /home/czhong/ScriptRepo/GetFastaSeqByID.pl --fasta=$pep_seq --id=$out_dir\/$decom[0]\/id_list --out=$out_dir\/$decom[0]\/pep.fasta";
  print "Compiling the list of recruited nucleotide sequences...\n";
  system "perl /home/czhong/ScriptRepo/GetFastaSeqByID.pl --fasta=$nuc_seq --id=$out_dir\/$decom[0]\/id_list --out=$out_dir\/$decom[0]\/nuc.fasta";
}
close $LIN,
